include/oneapi/dpl/pstl/parallel_backend_tbb.h

```diff
@@ -644,10 +644,13 @@ class __func_task : public __task
 
         assert(__parent != nullptr);
         assert(__parent->_M_refcount.load(::std::memory_order_relaxed) > 0);
-        if (--__parent->_M_refcount == 0)
+
+        auto __refcount = --__parent->_M_refcount;
+        __alloc.deallocate(this, *__ed);
```

Comment:
Is there a reason we have to decrement the refcount of the parent *before* we deallocate the child memory?

If not, couldn't we just move the deallocation call outside the `if` and simplify this code?

Reply:
@pavelkumbrasev,
Could you please answer that question?

Reply:
I believe it is some sort of an optimization (Used in several TBB patterns) where deallocation can be done concurrently  (the reference reached 0 and the thread can proceed with tree folding without waiting for task deallocation on another thread).
I think the benefit is really small so we can amend the optimization and use @danhoeflinger suggestion if it helps with readability.

Reply:
If it is faster to do it this way I wont object.  It may be worth adding a comment to make that clear going forward.  

The difference in readability is minor.  However, from someone unfamiliar with the code to start with, it makes it seem like there are more hidden connections which require this pattern.

Reply:
I've added a comment.